### DISCUSSION TITLE
Screen scaling on linux - Cursor restrained to the upper-left part of the screen after alt-tab

### DISCUSSION THREAD

**Discussion by Zyfarok** [ppy/osu#26352] [Q&A]
### Type

Game behaviour

### Bug description

I'm on Linux (Arch with KDE on Xorg)

I play on a 2560x1440 but I reduced the game to 1920x1080 using screen scaling (75%).
When I go back into the game after an alt-tab, sometimes (it's kinda random, I don't understand what's the trigger) my cursor gets locked in the upper-left part of the screen. I highlighted the approximate area in which my cursor can move in red in the screenshot bellow.

This only affects my mouse (my tablet can still move the cursor in the full window). I have raw input off (it does weird things if I enable it), and I disabled the integrated tablet driver to use my own OTD instance.
 
Relanching the game or disabling/re-enabling full-screen or changing the settings of the screen scaling seems to fix it temporarily but alt-tabbing again can just re-trigger it.

### Screenshots or videos

![osu_2023-12-10_19-30-57](https://github.com/ppy/osu/assets/10669609/21dda97e-b9bf-48f4-9d14-609947620272)

### Version

2023.1130.0-lazer

### Logs

[database.log](https://github.com/ppy/osu/files/13628854/database.log)
[input.log](https://github.com/ppy/osu/files/13628855/input.log)
[network.log](https://github.com/ppy/osu/files/13628856/network.log)
[performance.log](https://github.com/ppy/osu/files/13628857/performance.log)
[runtime.log](https://github.com/ppy/osu/files/13628858/runtime.log)

**Comment by Zyfarok** [comment]
Here is a small video to show what's happening and what I mean by random:
https://github.com/ppy/osu/assets/10669609/cc0e331c-81b8-4055-8262-82baa9b617a3
(does not load in my firefox somehow ? Works fine in chromium though)

(I use Alt+Tab to switch windows)

**Comment by bdach** [comment]
I don't think screen scaling should have anything to do with this? As implemented screen scaling does not confine the cursor to the scaled area. As in, your mouse can and will be positioned over the letterboxed area, it just won't show a pointer. Related: https://github.com/ppy/osu/issues/12098

Can you double-check and confirm whether screen scaling is related here in any way?

I'm not sure we can provide a fix either as this is probably an SDL issue or at least related to SDL, and on a linux system this would likely require replicating your exact WM setup to reproduce. But let's cross that bridge later, I want to confirm that screen scaling has nothing to do with this.

**Comment by Zyfarok** [comment]
> As implemented screen scaling does not confine the cursor to the scaled area.

It does on full-screen when you select "Scale everything". At least on my end, I can't exit the scaled area, even when that bug is not triggered. If that's not intentional then I don't know what's causing this.

> Can you double-check and confirm whether screen scaling is related here in any way?

I'll try tonight.

**Comment by bdach** [comment]
Okay, yeah, I apparently wasn't aware of the fact that it does confine in full-screen.

That said the linux caveats above still apply.

**Comment by Zyfarok** [comment]
I tested and can confirm this is only happening with screen scaling when selecting "Scale Everything".
I can always reach the top-left corner of the scaled screen, but the bottom-right-most position I can reach depends on the settings. Here is different examples of the bottom-right-most position I can reach for different settings:
![osu_2023-12-14_22-54-39](https://github.com/ppy/osu/assets/10669609/57900708-00ed-4524-aece-24854f1cd6be)
![osu_2023-12-14_23-01-19](https://github.com/ppy/osu/assets/10669609/45965854-c775-4def-a7b0-985f8c93e362)
![osu_2023-12-14_22-54-29](https://github.com/ppy/osu/assets/10669609/91be0229-a939-4aa2-a6b5-38379e8ff84b)
![osu_2023-12-14_23-01-41](https://github.com/ppy/osu/assets/10669609/3d2bdd28-8eba-45dc-9c99-fe958e3ca1d4)

My guess would be that it applies some incorrect scaling to the cursor position somehow, since the first 2 screens look like a scaling of each other and in the last two screenshots the position of the cursor is the same.

**Comment by Zyfarok** [comment]
Also, I just tried on Windows on the same machine with same settings and same monitors and it's not happening so it might be just linux.
But I know I had some slight issues with my main monitor which is that it advertises a 2160p TV resolution and I think both windows and linux were using it by default instead of the correct 1440p PC resolution. This might be related. I'll try to fiddle a bit with my xorg settings on linux and see if it changes anything.

From my calculation, it is as if the game was assuming that my screen has a resolution of ~1367x768 (+/- 1 px) when computing the bottom-right-most cursor position.

**Reply by Susko3** [reply]
If you resize the window, eg. to 1000x500, then trigger the bug in fullscreen, does the window size affect the calculation? I.e. will you get 1000x500 or 1367x768 from it?

It could be that the SDL confine rect is improperly applied to the windowed size and not the fullscreen size.

**Comment by dustywater** [comment]
Having the exact same issue when using screen scaling set to everything. Also using Arch Linux with KDE on X11 on a 1440p monitor.

**Comment by cihe13375** [comment]
Can reproduce on kubuntu 22.04 (kde, x11), on a 2160p monitor. 
1) screen resolution doesn't seem to matter; I changed my screen resolution to 720p and the bug still exists.
2) sometimes I can move the cursor to the bottom-left region after resuming from alt-tab, but once I move the cursor to the "confined" region at top-left it cannot go back anymore.
3) the bug disappears if both horizontal and vertical scalings are set to 100%.
4) enabling high precision mouse fixes it completely for me, and I didn't notice any issue caused by that.

**Comment by Zyfarok** [comment]
Would be interesting to test on Wayland to see if it's X11 specific. I'll check when I can.
